remill/Arch/X86/Semantics/MMX.h

```diff
@@ -73,6 +149,784 @@ DEF_SEM(PUNPCKHQDQ, D dst, S1 src1, S2 src2) {
+DEF_ISEL(PADDQ_MMXq_MEMq) = PADDQ<V64W, V64, MV64>;
+
+/*
+ * signed saturation arithematic for PADDS
+ * int8 res = a + b;
+ * int8 tmp = (res & ~(a | b)) < 0 ? 0x7f : res;
+ * int8 c = (~res & (a & b)) < 0 ? 0x80 : tmp;
+ *
+ */
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PADDSB, D dst, S1 src1, S2 src2) {
+	auto src1_vec = SReadV8(src1);
+	auto src2_vec = SReadV8(src2);
+	auto add_vec = SAddV8(src1_vec, src2_vec);
+	auto or_vec = SOrV8(src1_vec, src2_vec);
+	auto and_vec = SAndV8(src1_vec, src2_vec);
+
+	//Compute signed saturation arithematic on each bytes
+    auto dst_vec = SClearV8(SReadV8(dst));
+	for (size_t index = 0; index < NumVectorElems(add_vec); index++) {
+		auto add_elem = SExtractV8(add_vec, index);
+		auto or_elem = SExtractV8(or_vec, index);
+		auto and_elem = SExtractV8(and_vec, index);
+		auto tmp = Select(SCmpLt(SAnd(add_elem, SNot(or_elem)), decltype(add_elem)(0)), decltype(add_elem)(0x7F), add_elem);
+		auto value = Select(SCmpLt(SAnd(SNot(add_elem), and_elem), decltype(add_elem)(0)), decltype(add_elem)(0x80), tmp);
+		dst_vec = SInsertV8(dst_vec, index, value);
+	}
+	SWriteV8(dst, dst_vec);
+}
+
+DEF_ISEL(PADDSB_MMXq_MMXq) = PADDSB<V64W, V64, V64>;
+DEF_ISEL(PADDSB_MMXq_MEMq) = PADDSB<V64W, V64, MV64>;
+
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PADDSW, D dst, S1 src1, S2 src2) {
+	auto src1_vec = SReadV16(src1);
+	auto src2_vec = SReadV16(src2);
+	auto add_vec = SAddV16(src1_vec, src2_vec);
+	auto or_vec = SOrV16(src1_vec, src2_vec);
+	auto and_vec = SAndV16(src1_vec, src2_vec);
+
+	//Compute signed saturation arithematic on each bytes
+    auto dst_vec = SClearV16(SReadV16(dst));
+	for (size_t index = 0; index < NumVectorElems(add_vec); index++) {
+		auto add_elem = SExtractV16(add_vec, index);
+		auto or_elem = SExtractV16(or_vec, index);
+		auto and_elem = SExtractV16(and_vec, index);
+		auto tmp = Select(SCmpLt(SAnd(add_elem, SNot(or_elem)), decltype(add_elem)(0)), decltype(add_elem)(0x7FFF), add_elem);
+		auto value = Select(SCmpLt(SAnd(SNot(add_elem), and_elem), decltype(add_elem)(0)), decltype(add_elem)(0x8000), tmp);
+		dst_vec = SInsertV16(dst_vec, index, value);
+	}
+	SWriteV16(dst, dst_vec);
+}
+
+DEF_ISEL(PADDSW_MMXq_MMXq) = PADDSW<V64W, V64, V64>;
+DEF_ISEL(PADDSW_MMXq_MEMq) = PADDSW<V64W, V64, MV64>;
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PHADDW, D dst, S1 src1, S2 src2) {
+	auto lhs_vec = SReadV16(src1);
+	auto rhs_vec = SReadV16(src2);
+	auto dst_vec = SClearV16(SReadV16(dst));
+
+	// Compute the horizontal packing
+	auto vec_count = NumVectorElems(lhs_vec);
+	for(size_t index = 0; index < vec_count; index += 2) {
+		dst_vec = SInsertV16(dst_vec, index/2, SAdd(SExtractV16(lhs_vec, index), SExtractV16(lhs_vec, index+1)));
+	}
+	for(size_t index = 0; index < NumVectorElems(rhs_vec); index += 2) {
+		dst_vec = SInsertV16(dst_vec, (index + vec_count)/2, SAdd(SExtractV16(rhs_vec, index), SExtractV16(rhs_vec, index+1)));
```

Comment:
I am using (S,U)Add/Div function for computation on vector elements, Do you think it is good idea to use them for computing vector index as well?